src/libraries/System.Linq/src/System/Linq/First.cs

```diff
@@ -114,6 +114,19 @@ public static TSource FirstOrDefault<TSource>(this IEnumerable<TSource> source,
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.predicate);
             }
 
+            if (source.TryGetSpan(out ReadOnlySpan<TSource> span))
+            {
+                for (int i = 0; i < span.Length; i++)
+                {
+                    TSource element = span[i];
+                    if (predicate(element))
+                    {
+                        found = true;
+                        return element;
+                    }
+                }
+            }
+
             foreach (TSource element in source)
             {
                 if (predicate(element))
```

Comment:
I wonder if it makes sense to extract the `IEnumerable<T>` path to a separate method in cases like this to let the JIT inline the Span path and fold away the type checks.